### PR TITLE
Eliminate Lock Contention on Consumer Last Check-In Time

### PR DESCRIPTION
**Pull request by dgoodwin**
Problems surface in deployments with extremely high load. Because we update the consumer's row every time they authenticate, any request running long can cause other requests to wait longer than mysql's 50 second limit trying to get the lock to update that row.

To work around this I introduced a new table for consumer check-in times. The Consumer's last checkin time is calculated as the max in the table for that consumer.

A job is included to run periodically (every 8 hours by default) and clean out all but the most recent row for each consumer. In a really big deployment, this could be well into the millions. Multiple attempts were made to delete all but the most recent row for each consumer, none would scale to that level and work on all dbs. Instead we query checkins to be deleted in batches and delete them one chunk (30k) at a time until there are none left.

End result should eliminate lock wait timeouts on cp_consumer.lastcheckin because we're now doing fresh inserts in each request, rather than everyone trying to update the same row.

### CODE REVIEW

File: server/src/main/java/org/candlepin/model/CheckInCurator.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        // them until we stop finding results.
+        q.setMaxResults(DELETE_BATCH_SIZE);
+        List<String> checkInIds = q.list();
+
+        int totalDeleted = 0;
+        int totalQueries = 1;
+        while (checkInIds.size() > 0) {
+            log.debug("Found {} check-in IDs to be deleted.", checkInIds.size());
+            String deleteHql = "DELETE from CheckIn where id in (:deleteIds)";
+            Query deleteQuery = currentSession().createQuery(deleteHql);
```
Comment (Ceiu) [inline server/src/main/java/org/candlepin/model/CheckInCurator.java]: Minor optimization: Could move this query out of the loop to save some re-parsing time here (much like the select query above).

Comment (dgoodwin) [inline server/src/main/java/org/candlepin/model/CheckInCurator.java]: Good idea, on it.

File: server/src/main/java/org/candlepin/model/CheckIn.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+ * Checkins are added to this table, with older entries cleaned out after some configurable
+ * period of time. We're really only interested in the latest, but trying to update a
+ * single column was a frequent cause of lock contention in heavily loaded environments.
+ */
+@XmlRootElement
+@XmlAccessorType(XmlAccessType.PROPERTY)
+@Entity
+@Table(name = "cp_consumer_checkin")
+public class CheckIn extends AbstractHibernateObject {
+
```
Comment (wottop) [inline server/src/main/java/org/candlepin/model/CheckIn.java]: Any chance we need to be more specific, i.e. ConsumerCheckIn ?

Comment (dgoodwin) [inline server/src/main/java/org/candlepin/model/CheckIn.java]: Hmm, next up will be something like GuestIdCheckIn, I could go either way, could consider CheckIn just the most common ConsumerCheckIn, or we could be more specific. Think we should go specific?

**Comment by wottop** [comment]
After thinking about it, we can change it later just as easily as now.

ack

**Pull request merged**